File: No_2416/Sum of Prefix Scores of Strings.py
class Trie:
    def __init__(self):
        self.count = 0
        self.children = defaultdict()
        
    def insert(self, s):
        cur = self
        for c in s:
            if c not in cur.children:
                cur.children[c] = Trie()
            cur = cur.children[c]
            cur.count += 1
            
    def countScores(self, s):
        total = 0
        cur = self
        for c in s:
            cur = cur.children[c]
            total += cur.count
                
        return total
        

class Solution:
    def sumPrefixScores(self, words: List[str]) -> List[int]:
        t = Trie()
        for s in words:
            t.insert(s)
            
        ret = [t.countScores(s) for s in words]
        
        return ret
            
        

File: No_2416/Sum of Prefix Scores of Strings.cpp
class Trie {
    int count = 0;
    Trie* next[26] = {};

public:
    void insert(string& s) {
        Trie* cur = this;
        for (char c : s) {
            int i = c - 'a';
            if (cur->next[i] == nullptr)
                cur->next[i] = new Trie();
            cur = cur->next[i];
            cur->count++;
        }
    }
    
    int countScores(string& s) {
        int total = 0;
        Trie* cur = this;
        for (char c : s) {
            cur = cur->next[c-'a'];
            total += cur->count;
        }
        
        return total;
    }
};

class Solution {
public:
    vector<int> sumPrefixScores(vector<string>& words) {
        Trie t = Trie();
        for (string& s : words)
            t.insert(s);
        
        vector<int> ret;
        for (string& s : words)
            ret.push_back(t.countScores(s));
        
        return ret;
    }
};